### ISSUE TITLE
Finishing the object-oriented design of Tax-Calculator

### ISSUE THREAD

**Issue by martinholmer** [open-source-economics/Tax-Calculator#1720]
In pull request #1718, @MattHJensen asked:
> For the sake of developing a shared understanding of design objectives, could you describe the benefits you see of adopting these changes?  For instance, what is the value of "mov[ing] towards a more object oriented structure"?

Tax-Calculator has from the very beginning been developed using many of the principles of object-oriented programming (OOP).  But there are a few areas in which the current code does not follow OOP principles.  In recent weeks there has been a number of incidents that indicate a need to make Tax-Calculator comply more completely with OOP design principles.  In this issue, I describe the recent incidents and outline the changes that need to be made to complete the original OOP design of Tax-Calculator.

If you are not familiar with OOP, then you should do some background reading before continuing the read this issue.  A brief search of the Internet for "OOP in Python" produced [this document](https://www.python-course.eu/object_oriented_programming.php) as a reasonable first background document.

Tax-Calculator uses classes, class methods, data encapsulation, and even class inheritance (because the Policy, Behavior, Consumption, and Growdiff classes are all derived from the ParametersBase class), which are key concepts in OOP.  However, Tax-Calculator is weak on data hiding.   

Here is an excerpt from [another tutorial](https://www.python-course.eu/python3_object_oriented_programming.php) on OOP in Python:
> Data Abstraction, Data Encapsulation and Information Hiding are often synonymously used in books and tutorials on OOP. But there is a difference. Encapsulation is seen as the bundling of data with the methods that operate on that data. Information hiding on the other hand is the principle that some internal information or data is "hidden", so that it can't be accidentally changed. Data encapsulation via methods doesn't necessarily mean that the data is hidden. You might be capable of accessing and seeing the data anyway, but using the methods is recommended. Finally, data abstraction is present, if both data hiding and data encapsulation is used. This means data abstraction is the broader term: 
>
> Data Abstraction = Data Encapsulation + Data Hiding  

The OOP design principle of data abstraction has not been achieved in Tax-Calculator because we have no data hiding.  For additional discussion on the preferred approach to implementing data hiding in Python, see [this tutorial](https://www.python-course.eu/python3_properties.php).

And the current version of Tax-Calculator also deviates from OOP design principles when it defines many high-level utility functions that should be methods of the class the utility function is manipulating.

Before describing a plan to complete the OOP design of Tax-Calculator, consider a few reasons why this is important to do.  While OOP is not a silver bullet that solves all problems, it is a time-tested approach to developing complex computer programs, especially simulation models.  And OOP is widely used: for example, the Python programming language itself follows OOP design principles.

A few weeks ago, a new user (who was writing Python scripts that used the taxcalc package) asked why one of his scripts was crashing.  I looked at his script, responded with a revised script that did what he wanted to do, and then added this suggestion:
> Here are a couple of style rules I'd suggest to reduce the chance of having problems when doing Python programming with the taxcalc package:
>
> 1. Fully specify Consumption, Behavior, Records, and Policy objects before passing them to the Calculator class constructor.
>
> 2. After initializing a Calculator object, manipulate it using only Calculator class methods.
>
> Following these two rules means avoiding the manipulation of a Calculator object's private internal objects (like calc.records, calc.policy, calc.behavior, or calc.consumption).

And that user responded immediately saying:
> Thanks, this makes sense, and I'm glad to use the more OOP style you suggest.

So, our users (or, at least, some of them) understand OOP principles and are probably confused by the fact that Tax-Calculator uses OOP principles in many respects but not in others.

Another recent recognition of the need to refactor Tax-Calculator so it complies more throughly with OOP principles can be seen in the discussion of merged pull request #1708:
> [hd] do you think you should add an assertion statement [in utils.py]: assert isinstance(obj, Calculator)?
> [mh] Yes, I tried that, but we have never imported Calculator or Records into the utils.py file, so it's not possible [presumably because of problems caused by circular imports]
> [hd] Ok, I see.

But that conversation missed the fundamental solution, which is to make the high-level logic of that utility function a Calculator method, leaving the low-level logic in a revised utility function.

And a final reason for making Tax-Calculator adhere more closely to OOP principles is that it would make it far easier to show new users how to do Python programming with the taxcalc package and it would make Python programming with the taxcalc package easier for everybody (developers and users).  Consider the difference in this code before and after the changes in pull request #1718.

BEFORE CODE:
```
# income-tax distribution for 2018 with CLP and REF results side-by-side
dist_table1 = create_distribution_table(calc1,
                                        groupby='weighted_deciles',
                                        income_measure='expanded_income',
                                        result_type='weighted_sum')
assert isinstance(dist_table1, pd.DataFrame)
import numpy
if numpy.allclose(calc2.records.expanded_income,
                  calc1.records.expanded_income):
    income_measure = 'expanded_income'
else:
    income_measure = 'expanded_income_baseline'
    assert calc2.current_year == calc1.current_year
    assert calc2.records.dim == calc1.records.dim
    setattr(calc2.records, income_measure, getattr(calc1.records, 'expanded_income'))
dist_table2 = create_distribution_table(calc2,
                                        groupby='weighted_deciles',
                                        income_measure=income_measure,
                                        result_type='weighted_sum')
assert isinstance(dist_table2, pd.DataFrame)
dist_extract = pd.DataFrame()
dist_extract['funits(#m)'] = dist_table1['s006'] * 1e-6
dist_extract['itax1($b)'] = dist_table1['iitax'] * 1e-9
dist_extract['itax2($b)'] = dist_table2['iitax'] * 1e-9
dist_extract['aftertax_inc1($b)'] = dist_table1['aftertax_income'] * 1e-9
dist_extract['aftertax_inc2($b)'] = dist_table2['aftertax_income'] * 1e-9
```

AFTER CODE:
```
# income-tax distribution for 2018 with CLP and REF results side-by-side
dist_table1, dist_table2 = calc1.distribution_tables(calc2)
assert isinstance(dist_table1, pd.DataFrame)
assert isinstance(dist_table2, pd.DataFrame)
dist_extract = pd.DataFrame()
dist_extract['funits(#m)'] = dist_table1['s006'] * 1e-6
dist_extract['itax1($b)'] = dist_table1['iitax'] * 1e-9
dist_extract['itax2($b)'] = dist_table2['iitax'] * 1e-9
dist_extract['aftertax_inc1($b)'] = dist_table1['aftertax_income'] * 1e-9
dist_extract['aftertax_inc2($b)'] = dist_table2['aftertax_income'] * 1e-9
```

What needs to be done to complete the OOP design of Tax-Calculator?

1. Make the table utility functions into Calculator methods as done in (now merged) pull request #1718

2. Make the graph utility functions into Calculator methods as done in (now merged) pull request #1722

3. Make the `ce_aftertax_income` utility function into a Calculator method as done in (now merged) pull request #1723

4. Implement data hiding in the Calculator class by renaming the internal `policy`, `records`, `behavior`, and `consumption` objects as `__policy`, `__records`, `__behavior`, and `__consumption`.  See this discussion for details:
> Name | Notation | Behaviour
> -- | -- | --
> name | Public | Can be accessed from inside and outside
> _name | Protected | Like a public member, but they shouldn't be directly accessed from outside.
> __name | Private | Can't be seen and accessed from outside
>
> The following interactive sessions shows the behaviour of public, protected and private members:
>```
> >>> from encapsulation import Encapsulation
> >>> x = Encapsulation(11,13,17)
> >>> x.public
> 11
> >>> x._protected
> 13
> >>> x._protected = 23
> >>> x._protected
> 23
> >>> x.__private
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> AttributeError: 'Encapsulation' object has no attribute '__private'
> >>> 
>```

Pull requests #1753 and #1791 contain changes in pursuit of item 4 data hiding.

**Comment by MattHJensen** [comment]
Thanks very much for this explanation @martinholmer.

**Comment by MaxGhenis** [comment]
I'm not an expert in OOP, but it seems that Python isn't as OOP-oriented as, say, Java. For example, the consensus across a [couple](https://stackoverflow.com/q/2579840/1840471) [StackOverflow](https://stackoverflow.com/a/1568112/1840471) questions (and the [Python creator](https://stackoverflow.com/a/2579994/1840471)) is that Python should not use getters and setters, since attributes can be accessed directly.

This came up in #1791 (`getrecords()`) but thought it could make sense to discuss more generally here.

**Comment by martinholmer** [comment]
@MaxGhenis said about the new Calculator "getter" and "setter" methods:
> This came up in #1791 (getrecords()) but thought it could make sense to discuss more generally here.

@MaxGhenis thanks for pointing this out.  I've tried to improve things in pull request #1791.  What do you think about access methods in that pull request?

**Comment by MaxGhenis** [comment]
@martinholmer I see the [latest commit](https://github.com/martinholmer/Tax-Calculator/blob/9742300a185ccf73c00ea788e1ade3a96ec62530/taxcalc/calculate.py) doesn't have `get*` or `set*` methods. However, the `array` method is essentially to be a combination of `get` and `set`, so I don't think this addresses the issue.

As an analyst/user of taxcalc, some more thought on the topic makes me a bit skeptical of the new data hiding in general. Beginners will largely use taxcalc's built-in summary and visualization functions, but more advanced use cases involve accessing the `records` directly, and I'd guess experts at some point just think of taxcalc as a tool to build that `DataFrame`, from which they do their own pivoting and manipulation as needed. Hiding in this way suggests that this is to be avoided, which makes sense in OOP settings where the behind-the-scenes magic can change invisibly to the user, but I don't think applies to the `records` result which is a predictably-structured table.

There are exceptions, such as the `weights` attribute I initially encountered trouble with, and for this it makes sense to have a specific method to establish weights. This would align it as an input with raw records (CPS/PUF), and policy and behavioral params. But I'm not clear on how the new framework would have prevented my issue: I think I'd have done the same thing using either `array` or `__records`.

I realize you've already done a lot of work here, and I'm curious if I'm missing benefits on the dev side. But I see taxcalc's largest value as producing a high-quality set of aged records, so would be interested to know more rationale if the intent is to discourage their direct use.

**Comment by martinholmer** [comment]
@MaxGhenis, Thanks for [your thoughts](https://github.com/open-source-economics/Tax-Calculator/issues/1720#issuecomment-354509190) on issue 1720.  I appreciate the fact that you have spent the time to think about these issues.  Let me say a few things about your thoughts.

Let me assure you that the intent of the changes made in pursuit of #1720 is **not** to discourage the use of information in the Records object embedded in a Calculator object.  That is why several access methods have been added to the Calculator class.  

The main objective of all these changes is to make it easier and less error-prone to work with a Calculator object and its embedded objects, particularly the embedded Records object.  Python programmers should not be expected to understand the internal details of Tax-Calculator objects in order to write a useful Python script that uses the taxcalc package.  You are one of the most accomplished Python users of taxcalc, yet you don't fully understand (nor should you be expected to understand) the Records object.  You describe it this way:
> ... more advanced use cases involve accessing the  [Calculator object's]`records` directly, and I'd guess experts at some point just think of taxcalc as a tool to build that `DataFrame`, from which they do their own pivoting and manipulation as needed.

But actually a Records object is not a Pandas `DataFrame` and assuming it is would lead to programming errors.  (And these programming errors would most likely be reported as Tax-Calculator issues that would require time to resolve.)  Admittedly, a Records object has the "feel" of a Pandas `DataFrame`, and that is why the Calculator method `dataframe(variable_list)` has been added as part of the OOP redesign of the Calculator class.  This new method extracts the listed variables from the embedded Records object and returns then as a Pandas `DataFrame` object.  Then Python programmers can do anything they want with the returned `DataFrame`.

Other parts of your comment were helpful in that they stimulated further thinking and eventually a small, but important, improvement in the redesigned Calculator class.  So thanks again for your insightful comments.  If you are interested, this redesign was implemented in commit 7f574ca.

**Comment by martinholmer** [comment]
Issue #1720 has been resolved by the merge of pull request #1791.
